sharedbuffers/mapped_struct.py

```diff
@@ -320,6 +320,342 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
             raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
         return rv
 
+class mapped_dict(dict):
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        return mapped_list.pack_into((obj.keys(), [obj[k] for k in obj]), buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        key, values = mapped_list.unpack_from(buf, offs, idmap)
+        return cls(zip(key, values))
+
+def proxied_list_cmp(a, b):
+
+    alen = len(a)
+    blen = len(b)
+
+    if alen < blen:
+        return -1
+    elif alen > blen:
+        return 1
+
+    for i in xrange(alen):
+        selfe = a[i]
+        othere = b[i]
+
+        if selfe < othere:
+            return -1
+        elif selfe > othere:
+            return 1
+
+    return 0
+
+def islist(obj):
+    return isinstance(obj, (tuple, list, proxied_list, proxied_tuple))
+
+@cython.cclass
+class proxied_list(object):
+
+    cython.declare(
+        buf        = object,
+        pybuf      = 'Py_buffer',
+        offs       = cython.ulonglong,
+        idmap      = dict,
+    )
+
+    if cython.compiled:
+        def __del__(self):
```

Comment:
Did you check that this work correctly done conditional like this? The Cython compiler sometimes fails to understand dynamism when it comes to data or method definitions, and being a special method this, if Cython doesn't understand it, it won't work correctly (it's not a regular method, it needs to be registered as a finalizer in the type struct).

Also, did you check whether you should use `__del__` or `__dealloc__`? Because Cython's recommended finalizer for C data structures is the latter.

Reply:
I think that I have to use __dealloc__ if I use __cinit__ to initialize or allocate memory for the python object. But because I am using __init__, I can use __del__ to release the buffer.